From 1323214092853b9f81700c555cfc4259ae78d9e7 Mon Sep 17 00:00:00 2001
From: Dave Stevenson <dave.stevenson@raspberrypi.org>
Date: Mon, 30 Apr 2018 17:10:01 +0100
Subject: [PATCH 63/80] staging:vc04_services:bcm2835-codec: Expect framed
 data.

Switch to expecting the incoming encoded data to be framed
(ie one complete encoded frame per V4L2 buffer).
This helps kick the bitstream parser and avoids losing the last
couple of frames whilst it thinks that it isn't worth examining
the bitstream yet.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>
---
 .../vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c     | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c b/drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c
index ccd6a8add5a5..522bcd016b03 100644
--- a/drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c
+++ b/drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c
@@ -588,6 +588,15 @@ static void vb2_to_mmal_buffer(struct m2m_mmal_buffer *buf,
 	if (vb2->flags & V4L2_BUF_FLAG_KEYFRAME)
 		buf->mmal.mmal_flags |= MMAL_BUFFER_HEADER_FLAG_KEYFRAME;
 
+	/*
+	 * Adding this means that the data must be framed correctly as one frame
+	 * per buffer. The underlying decoder has no such requirement, but it
+	 * will reduce latency as the bistream parser will be kicked immediately
+	 * to parse the frame, rather than relying on its own heuristics for
+	 * when to wake up.
+	 */
+	buf->mmal.mmal_flags |= MMAL_BUFFER_HEADER_FLAG_FRAME_END;
+
 	buf->mmal.length = vb2->vb2_buf.planes[0].bytesused;
 	/*
 	 * Minor ambiguity in the V4L2 spec as to whether passing in a 0 length
-- 
2.17.0

